### ISSUE TITLE
Update GitHub webhook URLs 

### ISSUE THREAD

**Issue by hferentschik** [fabric8-services/fabric8-jenkins-proxy#94]
GitHub webhooks triggering Jenkins build have the form https://jenkins-<user\>-jenkins.8a09.starter-us-east-2.openshiftapps.com/github-webhook/

In order for the Jenkins Proxy to work, the webhook URL needs to be changed when the idling is enabled for a given user. 

The proxy should on startup iterate the list users and determine which user has the idler feature enabled. For these users, the webhook URL needs to be checked and potentially updated. 

The idea would be to delegate the change of the URL to the launchpad (generator-backend) which should already have the required code to set/modify the webhook. @gastaldi WDYT?

**Comment by gastaldi** [comment]
I think it's fine, as long as we don't add any specific logic related to other services in the launcher-backend. Is there a service we can call passing the authorization bearer token and the Jenkins webhook URL as in the format you described and return the modified URL (if Idler is enabled for the current user in this case)?

**Comment by hferentschik** [comment]
>  Is there a service we can call passing the authorization bearer token and the Jenkins webhook URL 

Well, I was hoping that you could be this service ;-) After all you have the code to create the webhook in the first place, right?

**Comment by gastaldi** [comment]
Yeah, we can create the webhook, but I'd prefer to keep the specific business logic separated (like checking if the user should use Idler or not) to avoid maintenance nightmares ;)

**Comment by vpavlin** [comment]
My idea would be to have 2 endpoints in backend:

1. Get list of existing webhook URLs

Now our code would check if the one we expect there is present

2. Update the existing webhook URL pointing to Jenkins with Proxy url

That would only mean backend needs to be able to list webhooks and update a single webhook - no logic hidden there

**Comment by vpavlin** [comment]
One note to this approach - this would work for current approach - enable specific users via toggle service, once we enable it for all, we need to switch all users to the proxy anyway, because looping through all of them and doing these checks would not make sense

**Comment by gastaldi** [comment]
@vpavlin isn't the webhook creation supposed to be performed when a quickstart is created? Or is it invoked from Jenkins or other system? We can definitely create the endpoints, I am just not sure who will call it and when

**Comment by vpavlin** [comment]
Yes, the webhook generation is handled by generator backend when codebase is generated, problem is we have existing tenants and we need to migrate them to new Jenkins route (jenkins-$USER-jenkins..openshift.com -> jenkins.openshift.io).

When we decide to enable Idler and Jenkins Proxy to all the users, new codebases should get the new route directly

**Comment by gastaldi** [comment]
I understand, but that doesn't really answer my question of who will call this endpoint. Is it going to be performed through some script and run manually by the SD team? 

I created https://github.com/fabric8-launcher/launcher-planning/issues/92 to track this work in the launcher-backend

**Comment by gastaldi** [comment]
Ah I think this answers it: 

>The proxy should on startup iterate the list users and determine which user has the idler feature enabled. For these users, the webhook URL needs to be checked and potentially updated.

**Comment by hferentschik** [comment]
> we need to switch all users to the proxy anyway, because looping through all of them and doing these checks would not make sense

Sure. I guess this comes also down to the level of integration we want to achieve. In my world, this is a toggle, so I should be able to turn this on/off as needed and everything is taken care of.  However, this is not the case atm. If the feature is turned on, there is manual interaction needed to change the existing webhook URLs (and vice versa). IMO, the change of the URL should be transparent to the user. 

So maybe instead of iterating all users on each startup, there is an API call one can make against the proxy to re-verify the URLs. This could then be triggered on changes to the feature toggle (there might  even be a way to register a callback for this).

@vpavlin, what is the strategy anyway? Once we want to enable this for all users, do we switch to the default activaton strategy (opposed to specifying user ids explicitly)?

**Comment by hferentschik** [comment]
> When we decide to enable Idler and Jenkins Proxy to all the users, new codebases should get the new route directly

Right, once Jenkins idler is the default, new tenants should be created with the right webhook directly. @gastaldi, where would one have to create an issue for that in order to track this dependency?

**Comment by gastaldi** [comment]
@hferentschik you can create issues for fabric8-launcher in https://github.com/fabric8-launcher/launcher-planning

**Comment by gastaldi** [comment]
@vpavlin what parameters you may provide when invoking the webhook endpoints? I'd need at least a Github token and the Github repository name.

**Comment by vpavlin** [comment]
So, we should be able to get OSIO tenant id and from there be able to find all his/her codebases..from my point of view it would be optimal if you could get GH token from auth service so that we minimise it's exposure between services.

**Comment by gastaldi** [comment]
To get the GH token, I'd need the Authorization Header to be set with the Bearer token. Is this something that I can expect to receive in this endpoint?

Just wondering if the Github REST API shouldn't be invoked directly instead

**Comment by vpavlin** [comment]
So how does the backend get the GH token when creating a repository?

**Comment by gastaldi** [comment]
It reads the `Authorization` header, then queries `auth.openshift.io` to grab the Github token

**Comment by vpavlin** [comment]
Ah, ok, so then Proxy could obtain the GH token and add Authorization header for Backend call

@alexeykazakov - any comments?

We'd like to call Launcher Backend to update GH webhook URL from Jenkins Proxy, so what should be the auth flow? Proxy obtains GH token with it's Auth token and passes it over to Launcher? Or should Launcher be able to obtain the GH token by itself?

**Comment by gastaldi** [comment]
If you have the GH token, you can access the Github REST API directly without any middle-man, just saying :)

**Comment by alexeykazakov** [comment]
@vpavlin I think the service which actually is going to use the GH token should obtain it from Auth. So, the flow is Proxy->Launcher->GitHub then it should be:
1. Proxy passes OSIO token to Launcher
2. Launcher calls Auth to obtain the GitHub token for this OSIO token
3. Launcher calls GitHub

**Comment by vpavlin** [comment]
> If you have the GH token, you can access the Github REST API directly without any middle-man, just saying :) See https://developer.github.com/v3/repos/hooks/

My issue with this is, that you already implement significant functionality around GH and the only thing proxy knows about Github is that it can send some specific payload from webhook and it should do something about it:)

So from the point of concern separation, Launcher is much more logical place for webhook update functionality. 

OTOH as this is a temporary solution until we are ok with enabling the proxy for everyone, it might make sense to implement this directly in Proxy and then drop it when it's not needed - what do you think, @hferentschik ?

**Comment by gastaldi** [comment]
If it's a temporary solution, I think it makes more sense to implement directly in Proxy, but I'll leave the decision up to you guys :)

**Comment by hferentschik** [comment]
> If you have the GH token, you can access the Github REST API directly without any middle-man, just saying :) See https://developer.github.com/v3/repos/hooks/

Sure. I guess the initial idea was that on the launcher side all the code would already be there and it would basically be a matter of exposing an API.

> If it's a temporary solution, I think it makes more sense to implement directly in Proxy,

Ok. Let's do this then for now. Let's see how this is panning out.

**Comment by maxandersen** [comment]
link to where launcher set the webhook (as mentioned by @gastaldi on chat): https://github.com/fabric8-launcher/launcher-backend/blob/master/addons/osio-addon/src/main/java/io/fabric8/forge/generator/kubernetes/CreateBuildConfigStep.java#L433

**Comment by hferentschik** [comment]
That looks like a hell of a method

**Comment by gastaldi** [comment]
Welcome to my world ;)

**Comment by hferentschik** [comment]
This relates to https://github.com/openshiftio/openshift.io/issues/1245 in OpenShift.io

**Comment by gastaldi** [comment]
I created a new endpoint that can be invoked with the following cURL command: 

```
curl 'https://forge.api.prod-preview.openshift.io/api/service/github/update-github-webhook?repository=repositoryName&url=https%3A%2F%2Fjenkins.openshift.io' -H 'Accept: application/json, text/plain, */*' -H 'Cache-Control: no-cache' -H 'Authorization: Bearer XXX' -X POST
```

So to invoke it, one would only need to pass the repository and url parameters AND the `Authorization` header with the OSIO token as specified in https://github.com/fabric8-services/fabric8-jenkins-proxy/issues/94#issuecomment-356691225

**Comment by hferentschik** [comment]
@gastaldi 
> I created a new endpoint that can be invoked with the following cURL command:

Nice. I like that. To be clear we need to things to happen here and this endpoint will help with the first part. It will make it easier for sure. The two parts are:

1. Convert existing webhooks to the new format 
   Your new endpoint will help with that. We are in the process of writing a script which iterates all existing build configurations, extract the repositories they are referring to and then potentially change the webhook URL. That said, with the current endpoint you are offering, we would still need to talk to GitHub to determine whether the repository contains the old style webhook or not and only then trigger your API. And even generally, how does your endpoint handle the fact that there might be other webhooks which come from other integrations. Should the API not be something along the line:

        `curl 'https://forge.api.prod-preview.openshift.io/api/service/github/update-github-webhook?repository=repositoryName&oldUrl=foo&newUrl=bar'`
    In this case we pass the old and new webhook URL, so that you have a way to determine whether the old URL even exist and only if you have a match you do the update. This would be by far the easiest for us. Our job would then just be to determine the list of potentially affected repositories and your service would do the heavy lifting. WDYT?
1. Make sure that newly provisioned users get the new style URLs. This ties to the code you have been referring to -  https://github.com/fabric8-launcher/launcher-backend/blob/master/addons/osio-addon/src/main/java/io/fabric8/forge/generator/kubernetes/CreateBuildConfigStep.java#L433. 

Does that make sense?

**Comment by gastaldi** [comment]
About 1: I made the change as requested. 
For 2, AFAIK the current code [grabs the Jenkins URL from Openshift](https://github.com/fabric8-launcher/launcher-backend/blob/master/addons/osio-addon/src/main/java/io/fabric8/forge/generator/kubernetes/CreateBuildConfigStep.java#L364). Can we expect that to return the correct URL or that needs to have a fixed value?

**Comment by hferentschik** [comment]
@gastaldi - what about return values? I guess a HTTP status of 200 means everything ok and everything else indicates an error, right? Should we on top of this return a boolean of other form of status info to indicate whether a replacement has taken place?

**Comment by gastaldi** [comment]
Yes this API is in staging. I have changed to return 302 when a webhook is created and 404 if the webhook is not found. See https://github.com/fabric8-launcher/launcher-backend/blob/master/web/src/main/java/io/fabric8/launcher/web/api/UpdateGitHubWebHooksResource.java

**Comment by hferentschik** [comment]
Sweet. I am working on the parts which will make use of this API...

**Comment by gastaldi** [comment]
Cool, let me know how that works out for you

**Comment by hferentschik** [comment]
Scripts for migration are merged via pull request #123. Leaving this issue open until we actually ran the migration.
